--- a/debian/xl2tpd.init	2018-03-20 16:18:28.916109469 -0700
+++ b/debian/xl2tpd.init	2018-03-21 10:17:38.695406642 -0700
@@ -20,6 +20,8 @@ DESC=xl2tpd

 test -x $DAEMON || exit 0

+mkdir -p /var/run/xl2tpd
+
 . /lib/lsb/init-functions

 # Include xl2tpd defaults if available
@@ -35,7 +37,7 @@ case "$1" in
   start)
 	echo -n "Starting $DESC: "
 	test -d ${XL2TPD_RUN_DIR:-/var/run/xl2tpd} || mkdir -p ${XL2TPD_RUN_DIR:-/var/run/xl2tpd}
-	start-stop-daemon --start --quiet --pidfile $PIDFILE \
+	start-stop-daemon --start --oknodo --quiet --pidfile $PIDFILE \
 		--exec $DAEMON -- $DAEMON_OPTS
 	echo "$NAME."
 	;;
@@ -59,7 +61,7 @@ case "$1" in
 	start-stop-daemon --oknodo --stop --quiet --pidfile \
 		$PIDFILE --exec $DAEMON
 	sleep 1
-	start-stop-daemon --start --quiet --pidfile \
+	start-stop-daemon --start --oknodo --quiet --pidfile \
 		$PIDFILE --exec $DAEMON -- $DAEMON_OPTS
 	echo "$NAME."
 	;;
