--- freeradius-client-1.1.6/configure.in	2008-03-05 10:02:55.000000000 -0800
+++ freeradius-client-1.1.6.new/configure.in	2017-08-11 16:48:27.414624439 -0700
@@ -206,7 +206,7 @@ AC_REPLACE_FUNCS(strdup strerror strcase
 AC_CHECK_FUNCS(flock fcntl uname gethostname sysinfo getdomainname)
 AC_CHECK_FUNCS(stricmp random rand snprintf vsnprintf)
 
-if test "$ac_cv_func_uname" = 'yes'
+if test "$ac_cv_func_uname" = 'yes' && test "$cross_compiling" = 'no'
 then
 	AC_MSG_CHECKING([for field domainname in struct utsname])
 	AC_TRY_RUN([
