--- a/templates/service/ssh/node.def	2018-05-03 16:31:52.010763264 -0700
+++ b/templates/service/ssh/node.def	2018-05-03 16:32:06.714762938 -0700
@@ -1,8 +1,8 @@
 priority: 500 # After syslog and logins
 help: Secure SHell (SSH) protocol
-delete:sudo /usr/sbin/invoke-rc.d ssh stop 
+delete:sudo /usr/sbin/invoke-rc.d sshd stop 
        sudo sh -c "echo 'SSHD_OPTS=' > /etc/default/ssh"
 end: if [ -z "$VAR(port/@)" ]; then exit 0; fi
      STR="SSHD_OPTS=\"-p $VAR(port/@)\""
      sudo sh -c "echo '$STR' > /etc/default/ssh"
-     sudo /usr/sbin/invoke-rc.d ssh restart
+     sudo /usr/sbin/invoke-rc.d sshd restart
