--- a/scripts/vyatta-system-nameservers	2018-05-04 13:56:28.649738798 -0700
+++ b/scripts/vyatta-system-nameservers	2018-05-04 13:56:43.857916399 -0700
@@ -40,7 +40,7 @@ restart_ntp () {
     # restart ntp if ntp is configured
     if [ -f /etc/ntp.conf ] && grep -q "^server" /etc/ntp.conf; then
         sudo /opt/vyatta/sbin/vyatta_update_ntp.pl
-        /usr/sbin/invoke-rc.d ntp restart >&/dev/null
+        /usr/sbin/invoke-rc.d ntpd restart >&/dev/null
     fi
 }

--- a/templates/system/ntp/node.def	2018-05-04 13:25:53.880033041 -0700
+++ b/templates/system/ntp/node.def	2018-05-04 13:26:29.612446219 -0700
@@ -7,10 +7,10 @@ end:
      then
 	 if pgrep -f -u ntp /usr/sbin/ntpd > /dev/null
 	 then
-	     sudo /usr/sbin/invoke-rc.d ntp force-reload
+	     sudo /usr/sbin/invoke-rc.d ntpd force-reload
          else
-	     sudo /usr/sbin/invoke-rc.d ntp start
+	     sudo /usr/sbin/invoke-rc.d ntpd start
 	 fi
      else
-	 sudo /usr/sbin/invoke-rc.d ntp stop
+	 sudo /usr/sbin/invoke-rc.d ntpd stop
      fi
--- a/scripts/system/vyatta_update_resolv.pl	2018-05-04 13:59:40.539979039 -0700
+++ b/scripts/system/vyatta_update_resolv.pl	2018-05-04 14:00:16.648400439 -0700
@@ -233,7 +233,7 @@ if (($dhclient_script == 1) || ($config_
         # this corresponds to what is done in name-server/node.def as a fix for bug 1300
         if ($ntp_config == 1) {
             system("sudo /opt/vyatta/sbin/vyatta_update_ntp.pl --dhclient-script $dhclient_script");
-            my $cmd_ntp_restart = "if [ -f /etc/ntp.conf ] && grep -q '^server' /etc/ntp.conf; then /usr/sbin/invoke-rc.d ntp restart >&/dev/null; fi &";
+            my $cmd_ntp_restart = "if [ -f /etc/ntp.conf ] && grep -q '^server' /etc/ntp.conf; then /usr/sbin/invoke-rc.d ntpd restart >&/dev/null; fi &";
             system($cmd_ntp_restart);
         }
     }
