diff -rupN strongswan-5.5.3/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.c strongswan-5.5.3.patched/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.c
--- strongswan-5.5.3/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.c	2016-06-30 07:20:10.000000000 -0700
+++ strongswan-5.5.3.patched/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.c	2017-08-07 16:23:04.371257976 -0700
@@ -192,7 +192,7 @@ METHOD(diffie_hellman_t, destroy, void,
 /*
  * Described in header.
  */
-openssl_diffie_hellman_t *openssl_diffie_hellman_create(
+openssl_diffie_hellman_t *openssl_diffie_hellman_create_custom(
 							diffie_hellman_group_t group, chunk_t g, chunk_t p)
 {
 	private_openssl_diffie_hellman_t *this;
@@ -252,5 +252,11 @@ openssl_diffie_hellman_t *openssl_diffie
 	DBG2(DBG_LIB, "size of DH secret exponent: %d bits", BN_num_bits(privkey));
 	return &this->public;
 }
+openssl_diffie_hellman_t *openssl_diffie_hellman_create( diffie_hellman_group_t group)
+{
+	chunk_t g;
+	chunk_t p;
+	openssl_diffie_hellman_create_custom(group, g, p);
+}
 
 #endif /* OPENSSL_NO_DH */
diff -rupN strongswan-5.5.3/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.h strongswan-5.5.3.patched/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.h
--- strongswan-5.5.3/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.h	2016-04-22 13:01:35.000000000 -0700
+++ strongswan-5.5.3.patched/src/libstrongswan/plugins/openssl/openssl_diffie_hellman.h	2017-08-07 16:23:04.371257976 -0700
@@ -44,8 +44,10 @@ struct openssl_diffie_hellman_t {
  * @param p				custom prime, if MODP_CUSTOM
  * @return				openssl_diffie_hellman_t object, NULL if not supported
  */
-openssl_diffie_hellman_t *openssl_diffie_hellman_create(
+openssl_diffie_hellman_t *openssl_diffie_hellman_create_custom(
 							diffie_hellman_group_t group, chunk_t g, chunk_t p);
+openssl_diffie_hellman_t *openssl_diffie_hellman_create(
+							diffie_hellman_group_t group);
 
 #endif /** OPENSSL_DIFFIE_HELLMAN_H_ @}*/
 
diff -rupN strongswan-5.5.3/src/libstrongswan/plugins/openssl/openssl_plugin.c strongswan-5.5.3.patched/src/libstrongswan/plugins/openssl/openssl_plugin.c
--- strongswan-5.5.3/src/libstrongswan/plugins/openssl/openssl_plugin.c	2016-10-08 05:17:09.000000000 -0700
+++ strongswan-5.5.3.patched/src/libstrongswan/plugins/openssl/openssl_plugin.c	2017-08-07 16:23:04.371257976 -0700
@@ -463,6 +463,7 @@ METHOD(plugin_t, get_features, int,
 			PLUGIN_PROVIDE(DH, MODP_1024_BIT),
 			PLUGIN_PROVIDE(DH, MODP_1024_160),
 			PLUGIN_PROVIDE(DH, MODP_768_BIT),
+		PLUGIN_REGISTER(DH, openssl_diffie_hellman_create_custom),
 			PLUGIN_PROVIDE(DH, MODP_CUSTOM),
 #endif
 #ifndef OPENSSL_NO_RSA
